tests/subsys/emds/emds_flash/src/main.c

```diff
@@ -0,0 +1,699 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <ztest.h>
+#include <errno.h>
+#include <zephyr.h>
+#include <storage/flash_map.h>
+#include <drivers/flash.h>
+#include <sys/crc.h>
+#include <dk_buttons_and_leds.h>
+#include <../subsys/emds/emds_flash.h>
+
+#if defined(CONFIG_BT)
+#include <mpsl.h>
+#include <sdc.h>
+#endif
+
+#define ADDR_OFFS_MASK 0x0000FFFF
+
+struct {
+	const struct device *fd;
+	uint32_t offset;
+	uint32_t size;
+	uint32_t ate_idx_start;
+	uint32_t data_wra_offset;
+} m_test_fd;
+
+struct emds_storage {
+	struct emds_fs fs;
+};
+
+/* Allocation Table Entry */
+struct test_ate {
+	uint16_t id;       /* data id */
+	uint16_t offset;   /* data offset within sector */
+	uint16_t len;      /* data len within sector */
+	uint8_t crc8_data; /* crc8 check of the entry */
+	uint8_t crc8;      /* crc8 check of the entry */
+} __packed;
+
+static struct emds_fs ctx;
+uint16_t m_sec_size;
+uint16_t m_sec_cnt;
+const struct flash_area *m_fa;
+
+static int flash_block_cmp(uint32_t addr, const void *data, size_t len)
+{
+	const uint8_t *data8 = (const uint8_t *)data;
+	int rc;
+	size_t bytes_to_cmp, block_size;
+	uint8_t buf[32];
+
+	block_size = 32 & ~(4 - 1U);
+
+	while (len) {
+		bytes_to_cmp = MIN(block_size, len);
+		rc = flash_read(m_test_fd.fd, addr, buf, bytes_to_cmp);
+		if (rc) {
+			return rc;
+		}
+		rc = memcmp(data8, buf, bytes_to_cmp);
+		if (rc) {
+			return 1;
+		}
+
+		len -= bytes_to_cmp;
+		addr += bytes_to_cmp;
+		data8 += bytes_to_cmp;
+	}
+
+	return 0;
+}
+
+static int flash_cmp_const(uint32_t addr, uint8_t value, size_t len)
+{
+	int rc;
+	size_t bytes_to_cmp, block_size;
+	uint8_t cmp[32];
+
+	block_size = 32 & ~(4 - 1U);
+
+	(void)memset(cmp, value, block_size);
+	while (len) {
+		bytes_to_cmp = MIN(block_size, len);
+		rc = flash_block_cmp(addr, cmp, bytes_to_cmp);
+
+		if (rc) {
+			return rc;
+		}
+		len -= bytes_to_cmp;
+		addr += bytes_to_cmp;
+	}
+	return 0;
+}
+
+static inline size_t align_size(size_t len)
+{
+	return (len + (4 - 1U)) & ~(4 - 1U);
+}
+
+static int data_write(const void *data, size_t len)
+{
+	const uint8_t *data8 = (const uint8_t *)data;
+	int rc = 0;
+	off_t offset;
+	size_t blen;
+	size_t temp_len = len;
+	uint8_t buf[32];
+
+	if (!temp_len) {
+		/* Nothing to write, avoid changing the flash protection */
+		return 0;
+	}
+
+	offset = m_test_fd.offset;
+	offset += m_test_fd.data_wra_offset & ADDR_OFFS_MASK;
+
+	blen = temp_len & ~(4 - 1U);
+
+	/* Writes multiples of 4 bytes to flash */
+	if (blen > 0) {
+		rc = flash_write(m_test_fd.fd, offset, data8, blen);
+		if (rc) {
+			/* flash write error */
+			goto end;
+		}
+		temp_len -= blen;
+		offset += blen;
+		data8 += blen;
+	}
+
+	/* Writes remaining (len < 4) bytes to flash */
+	/* Should NEVER be used for ATE!? */
+	if (temp_len) {
+		memcpy(buf, data8, temp_len);
+		(void)memset(buf + temp_len, 0, 4 - temp_len);
```

Comment:
I really do not understand using casting to (void) logic. It is applied for `memset` but not for all and but not for `memcpy`. Is there any logic behind this?